### PR TITLE
Mongodb TLS

### PR DESCRIPTION
**Pull request by krish7919**

### CODE REVIEW

File: docs/server/source/cloud-deployment-templates/ca-installation.rst [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+Step 4: Create the Self-Signed CA
+---------------------------------
+
+You can create the CA using:
+
+    .. code:: bash
+        
+        ./easyrsa init-pki
+        
+        ./easyrsa build-ca
```
Comment (ttmc) [inline docs/server/source/cloud-deployment-templates/ca-installation.rst]: This command asked me many questions, including:
```text
Common Name (eg: your user, host, or server name) [Easy-RSA CA]:temptest.zzzabc123g.net
```

I wasn't sure what to put for the answer. We need to add some documentation to help people with that question.

File: docs/server/source/cloud-deployment-templates/workflow.rst [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+#. Log in the the MongoDB Cloud Manager and select the group that will monitor
+   and backup this cluster from the dropdown box.
+
+#. Go to Settings, Group Settings and copy the ``Agent Api Key``.
+
+
+Step 2.5: Other common configuration options
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+#. Name of the mongodb instance (``mdb-instance-*``)
```
Comment (ttmc) [inline docs/server/source/cloud-deployment-templates/workflow.rst]: Are these names (like `mdb-instance-0`) something that the node operator can choose arbitrarily, or are they something they have to coordinate with the existing nodes in the cluster?

File: docs/server/source/cloud-deployment-templates/workflow.rst [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+Step 2.4: Get the API keys from MongoDB Cloud Manager
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+This helps to monitor and backup the MongoDB cluster in your node.
+
+#. Log in the the MongoDB Cloud Manager and select the group that will monitor
+   and backup this cluster from the dropdown box.
+
+#. Go to Settings, Group Settings and copy the ``Agent Api Key``.
```
Comment (ttmc) [inline docs/server/source/cloud-deployment-templates/workflow.rst]: Is the `Agent Api Key` going to be different for each BigchaindB node within the cluster, or is there just one for the entire cluster?

**Comment by ttmc** [comment]
@krish7919 Some of my comments above got hidden by GitHub when I pushed some commits, but they're still comments/questions I'd like you to address: just click **Show outdated**.

**Comment by ttmc** [comment]
I don't understand why you need to change `extendedKeyUsage = serverAuth` to `extendedKeyUsage = serverAuth,clientAuth` (in the file `x509-types/server`) in some cases but not all (i.e. not when creating a client certificate).

**Comment by ttmc** [comment]
@krish7919 Is it true that there's only one self-signed CA per cluster (associated with the organization managing the cluster)? That was my understanding, but now I'm not sure.

**Comment by ttmc** [comment]
@krish7919 In the last step of the page about creating a server/member certificate, I gather that those settings are in the MongoDB configuration file. Should they edit some MongoDB configuration file (where?), or just note those values for future use (e.g. in some ConfigMap), or what?

**Comment by ttmc** [comment]
@krish7919 On the page about generating a server/member certificate, you wrote:

> "You can replace the common name (`mdb-instance-0` above) with any other name so long as the instance can verify that it is the hostname."

Is it okay if every node uses `mdb-instance-0` or does every node have to have a different common name? Telling them that they _can_ change it is very frustrating. Change it to what? Why?

File: docs/server/source/cloud-deployment-templates/server-tls-certificate.rst [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+(The managing organization should already have a self-signed CA.)
+
+If you are the admin of the managing organization's self-signed CA,
+then you can import the CSR and use Easy-RSA to sign it. For example:
+
+.. code:: bash
+        
+   ./easyrsa import-req mdb-instance-0.req mdb-instance-0
+
+   ./easyrsa --subject-alt-name=DNS:localhost,DNS:mdb-instance-0 sign-req server mdb-instance-0
```
Comment (ttmc) [inline docs/server/source/cloud-deployment-templates/server-tls-certificate.rst]: I got an error message when I did this:
```text
~/bdb-cluster-ca/easy-rsa-3.0.1/easyrsa3$ ./easyrsa --subject-alt-name=DNS:localhost,DNS:mdb-instance-0 sign-req server mdb-instance-0

Note: using Easy-RSA configuration from: ./vars

Easy-RSA error:

Unknown cert type 'server'
```

File: docs/server/source/cloud-deployment-templates/revoke-tls-certificate.rst [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+Step 1: Revoke a Certificate
+----------------------------
+
+Since we used Easy-RSA version 3 to
+:ref:`set up the CA <How to Set Up a Self-Signed Certificate Authority>`,
+we use it to revoke certificates too.
+
+Go to the following directory (associated with the self-signed CA):
+``.../bdb-cluster-ca/easy-rsa-3.0.1/easyrsa3``.
+You need to be aware of the base filename used to import the certificate, and
```
Comment (ttmc) [inline docs/server/source/cloud-deployment-templates/revoke-tls-certificate.rst]: We should give an example "base filename".

File: docs/server/source/cloud-deployment-templates/ca-installation.rst (hunk 1 above)
Comment (krish7919) [inline docs/server/source/cloud-deployment-templates/ca-installation.rst]: It basically prompts you if you want to change the default options specified in the env vars in the `vars` file.

File: docs/server/source/cloud-deployment-templates/workflow.rst (hunk 2 above)
Comment (krish7919) [inline docs/server/source/cloud-deployment-templates/workflow.rst]: They can be any unique string across the cluster and yes, can be chosen by the node operator.

My personal opinion is that they be supplied by the federation to enable easier log parsing, configuration management and consistency. And hence a form like `mdb-instance-*` helps here.

File: docs/server/source/cloud-deployment-templates/workflow.rst (hunk 3 above)
Comment (krish7919) [inline docs/server/source/cloud-deployment-templates/workflow.rst]: Once for the entire cluster in the current workflow as there is a single MongoDB backup copy for the cluster in Cloud Manager.
If different nodes want to maintain separate copy of backup, they can specify a different API key (from a different group or account in the Cloud Manager).

File: docs/server/source/cloud-deployment-templates/revoke-tls-certificate.rst (hunk 5 above)
Comment (krish7919) [inline docs/server/source/cloud-deployment-templates/revoke-tls-certificate.rst]: The base filename is the name used to import the certificate using the `easyrsa import-req` command. Will mention it in the docs.

**Comment by krish7919** [comment]
> @krish7919 I don't understand why they need to change extendedKeyUsage = serverAuth to extendedKeyUsage = serverAuth,clientAuth (in the file x509-types/server) in some cases but not all (i.e. not when creating a client certificate).

As discussed during the review on Wed, this is because server certificates are used for cluster members and client certificates are used by the mongodb driver.

The server certificates need access to change the database and perform admin operations, whereas a client certificate will be used to perform only authorized operations.

**Comment by krish7919** [comment]
> @krish7919 Is it true that there's only one self-signed CA per cluster (associated with the organization managing the cluster)? That was my understanding, but now I'm not sure.

That is still true. Why do you think otherwise?

**Comment by krish7919** [comment]
> @krish7919 In the last step of the page about creating a server/member certificate, I gather that those settings are in the MongoDB configuration file. Should they edit some MongoDB configuration file (where?), or just note those values for future use (e.g. in some ConfigMap), or what?

They should edit the MongoDB configuration file as per their deployment.

The ConfigMap/Secrets thing will come into picture when deploying on a k8s cluster, which is a work in progress.

**Comment by krish7919** [comment]
> @krish7919 On the page about generating a server/member certificate, you wrote:
> 
> "You can replace the common name (mdb-instance-0 above) with any other name so long as the instance can verify that it is the hostname."
> Is it okay if every node uses the common name mdb-instance-0 or does every node have to have a different common name? Telling them that they can change it is very frustrating. Change it to what? Why?

Every node will have a different common name as per the the usual PKI rules.
MongoDB requires the common name to be the hostname too.

It's a unique identifier, I do not think changing a string here to the hostname of the instance one is hosting on should be frustrating.

The basics of PKI aren't covered here. These are steps to deploy it.
I can add a couple of links where users can refresh their PKI knowledge/basics before proceeding.
Would that help?

File: docs/server/source/cloud-deployment-templates/server-tls-certificate.rst (hunk 4 above)
Comment (krish7919) [inline docs/server/source/cloud-deployment-templates/server-tls-certificate.rst]: I will look into this. It will take some time as I do not have a setup now.

File: docs/server/source/cloud-deployment-templates/ca-installation.rst [hunk 6]
```diff
@@ -1,110 +1,108 @@
...
-We customize the CA setup by creating a ``vars`` file based on the existing
-``vars.example`` file bundled in the package.
-
-The values for the ``Distinguished Name`` given below (country, province,
-city, org, email) are references and can be changed as per your
-deployment. Refer to the ``vars.example`` file for detailed variable descriptions.
+We customize the CA configuration by creating a ``vars`` file based
+on the existing ``vars.example`` file.
+The country, province, city, org and email values should be changed 
+to the values for your organization.
```
Comment (krish7919) [inline docs/server/source/cloud-deployment-templates/ca-installation.rst]: The country, province, city, org and email are part of the DN and I think it should be mentioned here that they are part of the DN.

File: docs/server/source/cloud-deployment-templates/ca-installation.rst [hunk 7]
```diff
@@ -1,110 +1,108 @@
...
-        echo 'set_var EASYRSA_REQ_COUNTRY "DE"' >> vars
-        echo 'set_var EASYRSA_REQ_PROVINCE "Brandenburg"' >> vars
-        echo 'set_var EASYRSA_REQ_CITY "Berlin"' >> vars
-        echo 'set_var EASYRSA_REQ_ORG "BigchainDB GmbH"' >> vars
-        echo 'set_var EASYRSA_REQ_EMAIL "dev@bigchaindb.com" >> vars
+   echo 'set_var EASYRSA_REQ_COUNTRY "DE"' >> vars
+   echo 'set_var EASYRSA_REQ_PROVINCE "Berlin"' >> vars
+   echo 'set_var EASYRSA_REQ_CITY "Berlin"' >> vars
+   echo 'set_var EASYRSA_REQ_ORG "BigchainDB GmbH"' >> vars
+   echo 'set_var EASYRSA_REQ_EMAIL "dev@bigchaindb.com"' >> vars
```
Comment (krish7919) [inline docs/server/source/cloud-deployment-templates/ca-installation.rst]: Good catch with the single quote!

File: docs/server/source/cloud-deployment-templates/ca-installation.rst [hunk 8]
```diff
@@ -1,110 +1,108 @@
...
+   ./easyrsa init-pki
         
-        ./easyrsa build-ca
+   ./easyrsa build-ca
 
 
 You will be asked to enter a password for encrypting the ``ca.key`` file.
 Make sure to securely store this password. If you lose it, you risk not being
-able to add or remove entities from your PKI infrastructure in the future.
+able to add or remove entities in the future.
```
Comment (krish7919) [inline docs/server/source/cloud-deployment-templates/ca-installation.rst]: Just `entities` is vague. What defines an entity?
And hence `entities from your PKI infrastructure` seems better, I think.

File: docs/server/source/cloud-deployment-templates/workflow.rst [hunk 9]
```diff
@@ -7,114 +7,102 @@ We are constantly improving them.
...
+☐ Get the BigchainDB public keys of all the other nodes in the cluster.
+That list of public keys is known as the BigchainDB "keyring."
 
-This helps you set up auth, monitoring and billing of your APIs
 
-TODO(Troy): How to setup an account?
-TODO(Krish): How to get all the credentials from the portal?
+☐ Ask the managing organization
+for the FQDN used to serve the BigchainDB APIs
+and for a copy of the associated SSL/TLS certificate.
```
Comment (krish7919) [inline docs/server/source/cloud-deployment-templates/workflow.rst]: Why is this required? We do not need to share keys for any certificates ever.
The other org can map the DNS from x.com to y.com using a CNAME record, and https will pick up the certificates for y.com and all works smoothly as before.

File: docs/server/source/cloud-deployment-templates/ca-installation.rst [hunk 10]
```diff
@@ -2,81 +2,30 @@ Set Up a Self-Signed Certificate Authority
 ==========================================
 
 This page enumerates the steps *we* use to set up a self-signed certificate authority (CA).
-Your organization might do it differently.
+This is something that only needs to be done once per cluster,
+by the organization managing the cluster, i.e. the CA is for the whole cluster.
+We use Easy-RSA.
```
Comment (krish7919) [inline docs/server/source/cloud-deployment-templates/ca-installation.rst]: It needs to be done once for CA, and every time a new node comes up.

Comment (krish7919) [inline docs/server/source/cloud-deployment-templates/ca-installation.rst]: That is one more reason why I had the directory names as `federation-ca` and `client-cert` and `server-cert`.

File: docs/server/source/cloud-deployment-templates/server-tls-certificate.rst [hunk 11]
```diff
@@ -1,86 +1,28 @@
-Generate Cluster Member Certificate for MongoDB
-===============================================
+How to Generate a Server Certificate for MongoDB
+================================================
```
Comment (krish7919) [inline docs/server/source/cloud-deployment-templates/server-tls-certificate.rst]: There are 2 ways about it:
MongoDB calls it a `cluster member` certificate, and usual PKI terminology is `server` certificate.
I think we can stick with the PKI one to be agnostic of MongoDB terminology.

File: docs/server/source/cloud-deployment-templates/client-tls-certificate.rst [hunk 12]
```diff
@@ -24,38 +24,42 @@ Then :ref:`install and configure Easy-RSA in that directory <How to Install & Co
...
+   ./easyrsa init-pki
 
-        ./easyrsa gen-req bdb-instance-0 nopass
+   ./easyrsa gen-req bdb-instance-0 nopass
 
 
-Step 3: Sign the Client Certificate
------------------------------------
+You should change ``bdb-instance-0`` to a value based on the client
+the certificate is for.
```
Comment (krish7919) [inline docs/server/source/cloud-deployment-templates/client-tls-certificate.rst]: `could` would be better as I have mentioned, I prefer that the names are assigned to the instances for consistency, and to easily parse logs, easy config, etc.

File: docs/server/source/cloud-deployment-templates/ca-installation.rst (hunk 1 above)
Comment (ttmc) [inline docs/server/source/cloud-deployment-templates/ca-installation.rst]: I didn't know how to answer that _specific_ question, i.e. "Common Name (eg: your user, host, or server name) [Easy-RSA CA]:" Is `Easy-RSA CA` a valid, reasonable answer?

File: docs/server/source/cloud-deployment-templates/workflow.rst (hunk 2 above)
Comment (ttmc) [inline docs/server/source/cloud-deployment-templates/workflow.rst]: Okay, so each node _does_ have to coordinate with the other nodes in the cluster, because their instance names must be one-of-a-kind in the cluster. I guess one solution would be for the managing organization to assign a unique number to each BigchainDB node in the cluster, e.g. 1, 2, 3, 17. I will make a small commit to clarify this.

File: docs/server/source/cloud-deployment-templates/ca-installation.rst (hunk 1 above)
Comment (krish7919) [inline docs/server/source/cloud-deployment-templates/ca-installation.rst]: Yup. It needs a unique string that cannot be used again for signing server or client certificates.
So the default, `Easy-RSA CA` is perfectly valid, and so are values like `IPDB Foundation CA`, `BigchainDB GmbH CA`, etc.
Note: The CA common name is not constrained by the host name check, and hence _can_ have spaces in them.

File: docs/server/source/cloud-deployment-templates/workflow.rst (hunk 2 above)
Comment (krish7919) [inline docs/server/source/cloud-deployment-templates/workflow.rst]: They have to coordinate amongst themselves for BigchainDB public keys, but for self-signed CA, the CA needs to sign the certificates for each node.

If 2 nodes use the same CN, the first one will get its certificate request signed, but the second one will flag an error as a CN needs to be unique.

Comment (ttmc) [inline docs/server/source/cloud-deployment-templates/workflow.rst]: I added the notes about uniquely naming each instance in the cluster in this commit: https://github.com/bigchaindb/bigchaindb/pull/1456/commits/53bd121407ef67c1afb1cc622bcde6942259f528

File: docs/server/source/cloud-deployment-templates/ca-installation.rst (hunk 1 above)
Comment (ttmc) [inline docs/server/source/cloud-deployment-templates/ca-installation.rst]: Thanks, I added a comment about that in commit https://github.com/bigchaindb/bigchaindb/pull/1456/commits/6cfb461264b38c074659b34d4c07c60b83bd4050

Done.

File: docs/server/source/cloud-deployment-templates/workflow.rst (hunk 3 above)
Comment (ttmc) [inline docs/server/source/cloud-deployment-templates/workflow.rst]: I added a comment about one `Agent Api Key` per Cloud Manager backup in commit https://github.com/bigchaindb/bigchaindb/pull/1456/commits/810fdbb3703d8f340bbddf98c91d81fd2c73eeea 

Done.

File: docs/server/source/cloud-deployment-templates/workflow.rst [hunk 13]
```diff
@@ -38,8 +38,14 @@ and buy an SSL/TLS certificate for the FQDN.
...
-☐ Decide on, or find out some names of the things in your node:
-
+☐ Every MongoDB instance in the cluster must have a unique (one-of-a-kind) name.
+Ask the organization managing your cluster if they have a standard
+way of naming instances in the cluster.
+For example, maybe they assign a unique number to each node,
+so that if you're operating node 12, your MongoDB instance would be named
+``mdb-instance-12``.
+Similarly, other instances must also have unique names in the cluster.
+
```
Comment (krish7919) [inline docs/server/source/cloud-deployment-templates/workflow.rst]: Might want to add here:
`This MongoDB name must also be a resolvable name. The scope of how to achieve name resolution across your cluster and federation is beyond the scope of this document.`

Comment (ttmc) [inline docs/server/source/cloud-deployment-templates/workflow.rst]: These "names" are used in a variety of ways: as (part of) DNS names, Deployment names, StatefulSet names, etc. I don't want to get into that here. We can get into the details of making a DNS name resolvable elsewhere.

Comment (krish7919) [inline docs/server/source/cloud-deployment-templates/workflow.rst]: Alright.

File: docs/server/source/cloud-deployment-templates/ca-installation.rst (hunk 6 above)
Comment (ttmc) [inline docs/server/source/cloud-deployment-templates/ca-installation.rst]: DN?

File: docs/server/source/cloud-deployment-templates/ca-installation.rst (hunk 8 above)
Comment (ttmc) [inline docs/server/source/cloud-deployment-templates/ca-installation.rst]: Okay I added it back in commit https://github.com/bigchaindb/bigchaindb/pull/1456/commits/5c446e3dca31ed7b800d3493c9a6821c9ebe9b8b

File: docs/server/source/cloud-deployment-templates/workflow.rst (hunk 9 above)
Comment (ttmc) [inline docs/server/source/cloud-deployment-templates/workflow.rst]: Okay, in commit https://github.com/bigchaindb/bigchaindb/pull/1456/commits/06ae039e347053410dcc15b90949b6d6d6b045df I removed the bit about sharing SSL certificates (not how we will recommend doing it), and I added another step:

```text
☐ Ask the managing organization
for the FQDN used to serve the BigchainDB APIs
(e.g. ``api.orgname.net`` or ``bdb.clustername.com``).


☐ Make up an FQDN for your BigchainDB node (e.g. ``mynode.mycorp.com``).
Make sure you've registered the associated domain name (e.g. ``mycorp.com``),
and have an SSL certificate for the FQDN.
(You can get an SSL certificate from any SSL certificate provider).
```

File: docs/server/source/cloud-deployment-templates/ca-installation.rst (hunk 10 above)
Comment (ttmc) [inline docs/server/source/cloud-deployment-templates/ca-installation.rst]: The instructions on the `ca-installation.rst` page, to set up a self-signed CA, only have to be done once per cluster, right? There's only one CA per cluster. I don't understand why you wrote, "and everytime a new `bigchaindb node` is set up." (above)

I understand why there needs to be a different Easy-RSA directory for each of the three possible use cases.

File: docs/server/source/cloud-deployment-templates/server-tls-certificate.rst (hunk 11 above)
Comment (ttmc) [inline docs/server/source/cloud-deployment-templates/server-tls-certificate.rst]: :+1: The opening paragraph of that page explains how the MongoDB documentation calls server certificates "member certificates" so readers will understand.

File: docs/server/source/cloud-deployment-templates/client-tls-certificate.rst (hunk 12 above)
Comment (ttmc) [inline docs/server/source/cloud-deployment-templates/client-tls-certificate.rst]: I prefer "should" because the word "could" will make them wonder if they should, which means we would have to give them a bunch more information about how to decide. Why not just stick with a convention of giving every instance of everything in the cluster its own unique name, even if that's not always strictly necessary?

File: docs/server/source/cloud-deployment-templates/ca-installation.rst (hunk 6 above)
Comment (krish7919) [inline docs/server/source/cloud-deployment-templates/ca-installation.rst]: [Distinguished Name](https://en.wikipedia.org/wiki/X.509#Certificates).

File: docs/server/source/cloud-deployment-templates/ca-installation.rst (hunk 10 above)
Comment (krish7919) [inline docs/server/source/cloud-deployment-templates/ca-installation.rst]: Installation is once per cluster. Please disregard the earlier comment on installation.
I got the signing and installation process mixed up.

File: docs/server/source/cloud-deployment-templates/client-tls-certificate.rst (hunk 12 above)
Comment (krish7919) [inline docs/server/source/cloud-deployment-templates/client-tls-certificate.rst]: I get the point. I agree too now.

File: docs/server/source/cloud-deployment-templates/ca-installation.rst (hunk 6 above)
Comment (ttmc) [inline docs/server/source/cloud-deployment-templates/ca-installation.rst]: Okay I added a note about those things being part of the DN in commit https://github.com/bigchaindb/bigchaindb/pull/1456/commits/424839a96b84ecc4da81474a8f264db096c89c8e

File: docs/server/source/cloud-deployment-templates/ca-installation.rst (hunk 10 above)
Comment (ttmc) [inline docs/server/source/cloud-deployment-templates/ca-installation.rst]: :+1:

File: docs/server/source/cloud-deployment-templates/client-tls-certificate.rst (hunk 12 above)
Comment (ttmc) [inline docs/server/source/cloud-deployment-templates/client-tls-certificate.rst]: :+1:

File: docs/server/source/cloud-deployment-templates/revoke-tls-certificate.rst [hunk 14]
```diff
@@ -15,12 +15,13 @@ we use it to revoke certificates too.
...
-You need to be aware of the base filename used to import the certificate, and
-run:
+You need to be aware of the file name used to import the certificate using the
+``./easyrsa import-req`` before. Run the following command to revoke a
+certificate:
 
 .. code:: bash
 
-   ./easyrsa revoke <filename_base>
+   ./easyrsa revoke <filename>
```
Comment (ttmc) [inline docs/server/source/cloud-deployment-templates/revoke-tls-certificate.rst]: We should give an example filename (e.g. `mdb-instance-0.req`).

File: docs/server/source/cloud-deployment-templates/server-tls-certificate.rst (hunk 4 above)
Comment (ttmc) [inline docs/server/source/cloud-deployment-templates/server-tls-certificate.rst]: This problem was resolved by @krish7919 in commit https://github.com/bigchaindb/bigchaindb/pull/1456/commits/6abe1ead34fa67413e66d25c4104a03d3af1d936 (i.e. by removing `set_var EASYRSA_EXT_DIR "$EASYRSA/x509-types"` from the vars file).

:+1:
